botorch/acquisition/hentropy_search.py

```diff
@@ -0,0 +1,277 @@
+        )
+
+    @t_batch_mode_transform()
+    def forward(self, X: Tensor, A: Tensor) -> Tensor:
+        r"""Evaluate qHEntropySearch objective (q-HES) on the candidate set `X`.
+
+        Args:
+            X: Design tensor of shape `(batch) x q x num_dim_design`.
+            A: Action tensor of shape `(batch) x n_fantasy_at_design_pts
+                x num_actions x num_dim_action`.
+
+        Returns:
+            A Tensor of shape `(batch)`.
+        """
+
+        # construct the fantasy model of shape `n_fantasy_at_design_pts x b`
+        fantasy_model = self.model.fantasize(
+            X=X, sampler=self.design_sampler, observation_noise=True
+        )
+
+        # Permute shape of A to work with self.model.posterior correctly
+        A = A.permute(1, 0, 2, 3)
+
+        fantasized_outcome = self.action_sampler(fantasy_model.posterior(A))
+        # >>> n_fantasy_at_action_pts x n_fantasy_at_design_pts
+        # ... x batch_size x num_actions x 1
+
+        fantasized_outcome = fantasized_outcome.squeeze(dim=-1)
+        # >>> n_fantasy_at_action_pts x n_fantasy_at_design_pts
+        # ... x batch_size x num_actions
+
+        values = self.loss_function(A=A, Y=fantasized_outcome)
+        # >>> n_fantasy_at_design_pts x batch_size
+
+        # return average over the fantasy samples
+        return values.mean(dim=0)
+
+    def get_augmented_q_batch_size(self, q: int) -> int:
+        r"""Get augmented q batch size for optimization.
+
+        Args:
+            q: The number of candidates to consider jointly.
+
+        Returns:
+            The augmented size for optimization.
+        """
+
+        return q + self.n_fantasy_at_design_pts
+
+    def extract_candidates(self, X_full: Tensor) -> Tensor:
+        r"""We only return X as the set of candidates post-optimization.
+
+        Args:
+            X_full: A `b x (q + num_fantasies) x d`-dim Tensor with `b`
+                t-batches of `q + num_fantasies` design points each.
+
+        Returns:
+            A `b x q x d`-dim Tensor with `b` t-batches of `q` design points each.
+        """
+        return X_full[..., : -self.n_fantasy_at_design_pts, :]
+
+
+class qLossFunctionTopK(nn.Module):
+    r"""Batch loss function for the task of finding top-K
+    relative to the values of the objective function."""
+
+    def __init__(self, dist_weight=1.0, dist_threshold=0.5) -> None:
+        r"""Batch loss function for the task of finding top-K
+        relative to the values of the objective function.
+
+        Args:
+            loss_function_hyperparameters: hyperparameters for the loss function class.
+        """
+
+        super().__init__()
+        self.dist_weight = dist_weight
+        self.dist_threshold = dist_threshold
+
+    def forward(self, A: Tensor, Y: Tensor) -> Tensor:
+        r"""Evaluate batch loss function on a tensor of actions.
+
+        Args:
+            A: Actor tensor with shape `n_fantasy_at_design_pts x batch_size
```

Comment:
Nit: Inconsistency: Used "Action" tensor above but "Actor" tensor here - let's standardize this throughout the code.

Reply:
Also used `num_dim_action` and `action_dim` for the same thing